### ISSUE TITLE
Stuff to do for performance/build timeout 

### ISSUE THREAD

**Issue by easherma** [cityofaustin/techstack#3754]
These are _kinda_ in order, but not necessarily. 

1. implement PG Bouncer for connection pooling/management
   * http://www.pgbouncer.org/features.html
   * this would give us more control over queuing and how we handle open connections
1. async wokers on gunicorn
    * this would allow those requests to be concurrent on joplin, but 
    * during site builds, joplin seems to spend more time with requests waiting in a queue than anything else. this is probably because the site build process files off a bunch of requests at once, and gunicorn is handling them appropriately for how we have it set up but not really appropriate for the kind of app we've actually built. The short answer is that is what the above PG Bouncer is for
    * switching to this kind of setup was pretty easy (I have a PR out), but a naive implementation leads to lots of DB connections being created for every request
1. try a switch from gunicorn to uswgi
    * research seems kinda conflicting, but it seems uswgi is more robust and may deal with our variable load better. it also seems easier to configure by environment which might be nice. 
1. note pros and cons of vanilla gunicorn vs async gunicorn vs uswgi, make a decision ( for instance, if we stick with gunicorn it is designed to be run with a proxy buffer so we ought to implement that (there may be a heroku buildpack that saves some time with that)) 

1. upgrade PG in heroku to something being the free tier
    * this would allow more connections, and allow the PG database to have its own cache, which should greatly help response time of multiple requests 
1. Optimize our api queries to reduce the overall number or requests and improve response time
    * we should do this regardless of the above steps, but it will be more time intensive
    * note the outline on generally how querysets affect performance, links to relevant libraries: 
    https://github.com/cityofaustin/techstack/issues/3596

**Comment by desigonz** [comment]
Number 5 seems doable right away (and makes sense given the scale we're approaching). I gave Marni and Gerald a heads up about the upgrade so we're good to go, and I made an issue for it: #3756

**Comment by easherma** [comment]
This is true, and would be good to try just to see how it affects performance and response time. 
However, some sort of connection pooling is still also going to be desirable regardless. Happy to talk more about it in person too!

**Comment by easherma** [comment]
Doing more research, and both uswgi and gunicorn seem to expect to be run behind a reverse proxy. Since that means extra potential setup for uswgi and not a quick switch, it probably isn't worth comparing the two. 

@Niicck suggested avoiding a pgBouncer implementation unless absolutely needed, so I'm trying some python libraries to facilitate connection pooling in Django as part of joplin.

**Comment by easherma** [comment]
Useful info here:
https://github.com/benoitc/gunicorn/issues/996#issuecomment-93301359

**Comment by easherma** [comment]
https://scoutapm.com/shares/c4a76ff2-cfe0-49d5-a1e7-e401ab485239

**Comment by easherma** [comment]
https://scoutapm.com/shares/7f292a72-35bb-44a9-8ec9-5528a725f99e

**Comment by easherma** [comment]
Need to flesh out/out line notes more, but the Tl;DR

Neither upgrading the database nor switching the server methods will help improve the performance of building/publishing significantly. Fetching routes is usually 10-20 seconds, the server response rarely takes longer than 3 seconds even on a hobby dyno.

In contrast, we spend much longer actually bundling the app, and exporting the routes. Exporting the routes is the part that scales linearly, that incremental builds will help with. But it's not going to give us a big speed boost any time soon. ( Another build time source is likely the spinning up of the container we run the build in. )

So what causes timeouts? Well, gunicorn relies heavily on the OS CPU to handle queuing. So if there are 'lots' of requests, a smaller CPU doesn't seem to be able to handle managing them as well. The timeout appears to be the machine, essentially, and not the application, and not the database.

This is why upping the dynos made the problem go away. 

Using async or gthread workers would probably still help this, but its hard for me to test because we don't have a reliable way to reproduce a timeout.

**Comment by easherma** [comment]
Also/additional to the above, we could try migrating to a different service from gunicorn entirely, there are some promising ones on this list in terms of performance:
https://www.appdynamics.com/blog/engineering/a-performance-analysis-of-python-wsgi-servers-part-2/

**Comment by easherma** [comment]
From the research I've done, in the future we might want to migrate away from patching in gevent as it may affect other parts of the app and in general be hard to debug unintended consequences. We'll have to see if we run into any of those.

Once publisher is fully hooked into the process, time-outs due to builds should be less of a problem, so we can probably migrate back to a more simple gunicorn setup.

Or if we want something more sustainable, it would behoove us to either set up nginx as a reverse proxy to handle bursts of traffic or add pgbouncer as a way of handiling connections. Nginx setup would also help prevent DDOS attacks, though Heroku is maddeningly unclear on if they already provide a reverse proxy for apps you run on their server.
